src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/FindUnwantedIslandsTask.cpp

```diff
@@ -0,0 +1,136 @@
+    stateInfo.setProgress(0);
+    int progressZeroPoint = startNucleotideNumber;
+    while (leftNucleotide /*+ ISLAND_LENGTH*/ <= rightNucleotide) {
+        U2Region islandCandidate(leftNucleotide, ISLAND_LENGTH);
+        bool isIsland = hasUnwantedConnections(islandCandidate);
+        if (isIsland) {
+            //The obvious limit - we don't need regions which couldn't fit the primer
+            if (lengthBetweenIslands != 0/*>= overlap.minValue*/) {
+                U2Region newRegion(startNucleotideNumber, lengthBetweenIslands);
+                text2LogAboutFoundRegion(newRegion);
+                regionsBetweenIslands << newRegion;
+            }
+            startNucleotideNumber = leftNucleotide;
+            lengthBetweenIslands = 0;
+        } else {
+            lengthBetweenIslands++;
+        }
+        leftNucleotide++;
+        stateInfo.setProgress(100 - (double(rightNucleotide - leftNucleotide) / double(rightNucleotide - progressZeroPoint)) * 100);
+    }
+    stateInfo.setProgress(100);
+    U2Region newRegion(startNucleotideNumber, lengthBetweenIslands);
+    text2LogAboutFoundRegion(newRegion);
+    regionsBetweenIslands << U2Region(startNucleotideNumber, lengthBetweenIslands);
+
+    if (!regionsBetweenIslands.isEmpty()) {
+        QString regions;
+        for (const auto& region : regionsBetweenIslands) {
+            regions += QString("%1,").arg(regionToString(region));
+        }
+        regions = regions.left(regions.size() - 1);
+        taskLog.details(tr("The following regions are located between unwanted islands: %1").arg(regions));
+    } else {
+        taskLog.details(tr("The whole region is filled with unwanted islands, no regions between them has been found"));
+    }
+
+    // Sort sequence FROM the amplified fragment
+    std::sort(regionsBetweenIslands.begin(), regionsBetweenIslands.end(),
+        [](const U2Region& first, const U2Region& second) -> bool {
+        return first.startPos > second.startPos;
+    });
+}
+
+const QList<U2Region>& FindUnwantedIslandsTask::getRegionBetweenIslands() const {
+    return regionsBetweenIslands;
+}
+
+bool FindUnwantedIslandsTask::hasUnwantedConnections(const U2Region& region) const {
+    /**
+     * Sequence to find the unwanted connections in.
+     */
+    QByteArray regionSequence = sequence.mid(region.startPos, region.length);
+    /**
+     * It's reverse complement representation.
+     */
+    //QByteArray revComRegionSequence = DNASequenceUtils::reverseComplement(regionSequence);
+    bool isUnwantedSelfDimer = UnwantedConnectionsUtils::isUnwantedSelfDimer(regionSequence,
+                                                                             UNWANTED_DELTA_G,
+                                                                             UNWANTED_MELTING_TEMPERATURE,
+                                                                             UNWANTED_MAX_LENGTH);
+
+    //TODO: hairpins
+    //TODO: find out if hetero-dimers are required
```

Comment:
Is the algorithm complete for use in production?

Reply:
Not hairpins